### ISSUE TITLE
Question about Gradients(Test)

### ISSUE THREAD

**Issue by rikvdkleij** [tensorflow/haskell#222]
To support the `tanh` activation function I have implemented the `opgrad "Tanh"` and `opgrad "Tanhgrad"`  functions similar as the Relu ones,
It's implemented in this commit https://github.com/active-group/haskell/commit/f3889b9573ee73465660462c6ea3031ab58d99df
Eventually I want to create PR.

In practise the `tanh` function seems to work properly.
Now I want to implement the tests (similar as the Relu ones) in `GradientTest.hs`. But I do not understand the  working of the `gradients` function in `Gradient.hs`.

To start with, I have created this test:
```
testTanhGrad :: Test
testTanhGrad = testCase "testTanhGrad" $ do
    [dx] <- TF.runSession $ do
        x <- TF.render $ TF.vector [1 :: Float]
        let y = TF.tanh x
        TF.gradients y [x] >>= TF.run
    V.fromList [0] @=? dx
```
which always returns 0 independently of the `x` value. I noticed similar behavior for `testReluGrad`.

Can you help me in understanding the gradients function (tests)?

**Comment by rikvdkleij** [comment]
After digging further I noticed this. 

The arguments of a TF unary gradient function have to be in this order:
https://github.com/tensorflow/tensorflow/blob/9590c4c32dd4346ea5c35673336f5912c6072bf2/tensorflow/core/ops/math_ops.cc#L167

I had used the same order as the `relu` function but it seems to be that case for `tanh` and also for the `sigmoid` function.
So here https://github.com/active-group/haskell/commit/f3889b9573ee73465660462c6ea3031ab58d99df#diff-b01d07da7932f40f70a55621563e020eR463
`tanhGrad dz x` has to become `tanhGrad x dz`.

Then when I run the test again an input value of 0 gives 1 which seems to be okay. But fore example 10 gives -99.

**Comment by fkm3** [comment]
> To start with, I have created this test:
> ...
> which always returns 0 independently of the x value. I noticed similar behavior for testReluGrad.

What you are seeing for `relu` is probably expected. `testReluGrad` should see 0 for all negative inputs and 1 for all positive inputs. `testReluGradGrad` should see for 0 for all inputs (`relu` doesn't have a second derivative).

> The arguments of a TF unary gradient function have to be in this order:
https://github.com/tensorflow/tensorflow/blob/9590c4c32dd4346ea5c35673336f5912c6072bf2/tensorflow/core/ops/math_ops.cc#L167

Good catch. You can see the argument order in the haskell docs:
https://tensorflow.github.io/haskell/haddock/tensorflow-core-ops-0.2.0.0/TensorFlow-GenOps-Core.html#v:tanhGrad
It is unfortunate that they are inconsistent.

I think the right implementation is

```
opGrad "Tanh" _ [toT -> x] [dz] = [Just $ CoreOps.tanhGrad (tanh x) dz]
```

`reluGrad` accepts the original argument `x`, but `tanhGrad` accepts the output `y = tanh(x)`.

> It's implemented in this commit active-group@f3889b9

FYI, the lines

```haskell
opGrad "TanhGrad" _ [_, toT -> x ] [dz] = [Just $ tanhGrad dz x, Just $ CoreOps.zerosLike x]

"TanhGrad" -> 1
```

are for the second derivative. You probably don't want it, so you can just delete. If you do want it, that implementation is incorrect.

**Comment by rikvdkleij** [comment]
@fkm3 Thanks for your help!

> What you are seeing for relu is probably expected. testReluGrad should see 0 for all negative inputs and 1 for all positive inputs. testReluGradGrad should see for 0 for all inputs (relu doesn't have a second derivative).

Yep, I realised that later on.

>  think the right implementation is
`opGrad "Tanh" _ [toT -> x] [dz] = [Just $ CoreOps.tanhGrad (tanh x) dz]`

This works! Thanks!
I understood from the comments that `tanh x` would be applied in the `tanhgrad`. 

I have removed the second derivative for `tanh` and will create PR soon 😄
